ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -606,7 +607,56 @@ genUpdateSchedule
   -> SecurityParam
   -> Gen (Schedule ChainUpdate)
 genUpdateSchedule updateBehavior securityParam =
-    genChainUpdates updateBehavior securityParam 10 >>= genSchedule
+    genChainUpdates updateBehavior securityParam 10
+      >>= genSchedule schedulingStrategy
+  where
+    schedulingStrategy = case behaviorValidity updateBehavior of
+        -- Otherwise, invalid behavior within a single tick might not be visible
```

Comment:
Another idea: De-atomize updating the chain with the list of `ChainUpdate`s, and rely on io-sim's guarantee that we will still notice all intermediate states.

Reply:
> Another idea: De-atomize updating the chain with the list of `ChainUpdate`s, and rely on io-sim's guarantee that we will still notice all intermediate states.

This does not seem to work directly, see an example here:

<details>
<summary>De-atomification</summary>

```diff
diff --git a/ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs b/ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs
index e49b191bf..28ebd5332 100644
--- a/ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs
+++ b/ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs
@@ -335,15 +335,15 @@ runChainSync securityParam (ClientUpdates clientUpdates)
 
         -- Server
         whenJust (Map.lookup tick (getSchedule serverUpdates)) $ \chainUpdates ->
-          atomically $ do
+          forM_ (toChainUpdates chainUpdates) $ \chainUpdate -> atomically $ do
             chainProducerState <- readTVar varChainProducerState
-            case CPS.applyChainUpdates
-                   (map (fmap TestHeader) (toChainUpdates chainUpdates))
+            case CPS.applyChainUpdate
+                   (TestHeader <$> chainUpdate)
                    chainProducerState of
               Just chainProducerState' ->
                 writeTVar varChainProducerState chainProducerState'
               Nothing                  ->
-                error $ "Invalid chainUpdates: " <> show chainUpdates <>
+                error $ "Invalid chainUpdate: " <> show chainUpdate <>
                         " for " <> show (chainState chainProducerState)
         atomically $ writeTVar varLastUpdate tick
 
@@ -619,7 +619,7 @@ genUpdateSchedule updateBehavior securityParam =
     schedulingStrategy = case behaviorValidity updateBehavior of
         -- Otherwise, invalid behavior within a single tick might not be visible
         -- to the client due to a rollback.
-        Invalid -> SingleItemPerTickStrategy
+        Invalid -> DefaultSchedulingStrategy
         Valid   -> DefaultSchedulingStrategy
 
 genInvalidBlocks ::
```
With e.g. `--quickcheck-replay=790408`, you get
```
    chainSync: FAIL (3.44s)
      *** Failed! Falsified (after 3576 tests and 16 shrinks):
      ChainSyncClientSetup:
      securityParam: 5
      clientUpdates:
      --
      serverUpdates:
      3: AddBlock (H:[3],S:1,B:1), AddBlock (H:[3,1],S:2,B:2), SwitchFork <- Origin -> (H:[0],S:1,B:1) (H:[0,0],S:2,B:2) (H:[0,0,2],S:3,B:3) (H:[0,0,2,0],S:4,B:4)
      --
      startTick: Tick {tickToWord64 = 1}
      invalidBlocks:
      1: [3]
      11: [1,0,1,2,0,1,1,1,0,4]


      Server behavior: InvalidChainBehavior
      Client chain: Origin ]
      Server chain: Origin ] (H:[0],S:1,B:1) :> (H:[0,0],S:2,B:2) :> (H:[0,0,2],S:3,B:3) :> (H:[0,0,2,0],S:4,B:4)
      Synced fragment: Origin ] (H:[0],S:1,B:1) :> (H:[0,0],S:2,B:2) :> (H:[0,0,2],S:3,B:3) :> (H:[0,0,2,0],S:4,B:4)
      Trace:
      1 | Protocol: Send (ClientAgency TokIdle,MsgFindIntersect [Origin,Origin])
      1 | Protocol: Recv (ServerAgency TokIntersect,MsgIntersectFound Origin TipGenesis)
      1 | Client: TraceFoundIntersection Origin TipGenesis TipGenesis
      1 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      1 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgRollBackward Origin TipGenesis)
      1 | Client: TraceRolledBack Origin
      1 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      1 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgAwaitReply)
      3 | Protocol: Recv (ServerAgency TokNext TokMustReply,MsgRollForward TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [3]), tbSlot = SlotNo 1, tbValid = Invalid, tbPayload = ()}} Tip (SlotNo 1) (testHashFromList [3]) (BlockNo 1))
      3 | Client: TraceDownloadedHeader (TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [3]), tbSlot = SlotNo 1, tbValid = Invalid, tbPayload = ()}})
      3 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      3 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgRollBackward Origin Tip (SlotNo 4) (testHashFromList [0,0,2,0]) (BlockNo 4))
      3 | Client: TraceRolledBack Origin
      3 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      3 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgRollForward TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0]), tbSlot = SlotNo 1, tbValid = Valid, tbPayload = ()}} Tip (SlotNo 4) (testHashFromList [0,0,2,0]) (BlockNo 4))
      3 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      3 | Client: TraceDownloadedHeader (TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0]), tbSlot = SlotNo 1, tbValid = Valid, tbPayload = ()}})
      3 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgRollForward TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0,0]), tbSlot = SlotNo 2, tbValid = Valid, tbPayload = ()}} Tip (SlotNo 4) (testHashFromList [0,0,2,0]) (BlockNo 4))
      3 | Client: TraceDownloadedHeader (TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0,0]), tbSlot = SlotNo 2, tbValid = Valid, tbPayload = ()}})
      3 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      3 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgRollForward TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0,0,2]), tbSlot = SlotNo 3, tbValid = Valid, tbPayload = ()}} Tip (SlotNo 4) (testHashFromList [0,0,2,0]) (BlockNo 4))
      3 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      3 | Client: TraceDownloadedHeader (TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0,0,2]), tbSlot = SlotNo 3, tbValid = Valid, tbPayload = ()}})
      3 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgRollForward TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0,0,2,0]), tbSlot = SlotNo 4, tbValid = Valid, tbPayload = ()}} Tip (SlotNo 4) (testHashFromList [0,0,2,0]) (BlockNo 4))
      3 | Client: TraceDownloadedHeader (TestHeader {testHeader = TestBlockWith {tbHash = (testHashFromList [0,0,2,0]), tbSlot = SlotNo 4, tbValid = Valid, tbPayload = ()}})
      3 | Protocol: Send (ClientAgency TokIdle,MsgRequestNext)
      3 | Protocol: Recv (ServerAgency TokNext TokCanAwait,MsgAwaitReply)

      Invalid behavior wasn't caught
```

</details>

In any case, I don't think that this restriction actually is a problem. Rather, it is a fundemantal property of testing that we disconnect from dishonest peers that we actually see their dishonest behavior -- just like in a real node, we can't punish something we don't notice, e.g. due to a timely rollback.